servicetalk-http-api/src/test/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategyTest.java

```diff
@@ -0,0 +1,207 @@
+            pubs.add(pub);
+            return pub;
+        }))).subscribe(subscriber);
+        subscriber.request(MAX_VALUE);
+    }
+
+    @Test
+    public void errorWithNoAddresses() throws Exception {
+        TestPublisher<ServiceDiscovererEvent<String>> sdEvents = pubs.take();
+        sdEvents = triggerRetry(sdEvents);
+        verifyNoEventsReceived();
+        sendUpAndVerifyReceive("addr1", sdEvents);
+    }
+
+    @Test
+    public void newAddressPostRetry() throws Exception {
+        TestPublisher<ServiceDiscovererEvent<String>> sdEvents = pubs.take();
+
+        final DefaultServiceDiscovererEvent<String> evt1 = sendUpAndVerifyReceive("addr1", sdEvents);
+
+        sdEvents = triggerRetry(sdEvents);
+
+        verifyNoEventsReceived();
+        final DefaultServiceDiscovererEvent<String> evt2 = new DefaultServiceDiscovererEvent<>("addr2", true);
+        sdEvents.onNext(evt2);
+
+        assertThat("Unexpected event received", subscriber.takeItems(),
+                containsInAnyOrder(flipAvailable(evt1), evt2));
+    }
+
+    @Test
+    public void overlapAddressPostRetry() throws Exception {
+        TestPublisher<ServiceDiscovererEvent<String>> sdEvents = pubs.take();
+
+        final DefaultServiceDiscovererEvent<String> evt1 = sendUpAndVerifyReceive("addr1", sdEvents);
+        final DefaultServiceDiscovererEvent<String> evt2 = sendUpAndVerifyReceive("addr2", sdEvents);
+
+        sdEvents = triggerRetry(sdEvents);
+
+        verifyNoEventsReceived();
+
+        sdEvents.onNext(evt1); // previously existing, should not be emitted
+        verifyNoEventsReceived();
+
+        final DefaultServiceDiscovererEvent<String> evt3 = new DefaultServiceDiscovererEvent<>("addr3", true);
+        sdEvents.onNext(evt3); // threshold breach, should evict addr2
+
+        assertThat("Unexpected event received", subscriber.takeItems(),
+                containsInAnyOrder(flipAvailable(evt2), evt3));
+    }
+
+    @Test
+    public void errorWhileRetaining() throws Exception {
+        TestPublisher<ServiceDiscovererEvent<String>> sdEvents = pubs.take();
+
+        final DefaultServiceDiscovererEvent<String> evt1 = sendUpAndVerifyReceive("addr1", sdEvents);
+        final DefaultServiceDiscovererEvent<String> evt2 = sendUpAndVerifyReceive("addr2", sdEvents);
+
+        sdEvents = triggerRetry(sdEvents);
+
+        verifyNoEventsReceived();
+
+        sdEvents.onNext(evt1); // previously existing, should not be emitted
+        verifyNoEventsReceived();
+
+        sdEvents = triggerRetry(sdEvents); // error while retaining
+
+        final DefaultServiceDiscovererEvent<String> evt3 = new DefaultServiceDiscovererEvent<>("addr3", true);
+        sdEvents.onNext(evt3); // threshold breach, should evict addr2
+
+        assertThat("Unexpected event received", subscriber.takeItems(),
+                containsInAnyOrder(flipAvailable(evt1), flipAvailable(evt2), evt3));
```

Comment:
The comment above says "should evict addr2", but `evt1` is also evicted. Can you please clarify why this is expected?